### PR TITLE
[DataGrid] Allow passing readonly arrays to `pageSizeOptions` prop

### PR DESCRIPTION
**Pull request by pcorpet**
Contributes to #10685

- [X] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

**Comment by pcorpet** [comment]
Will require https://github.com/mui/material-ui/pull/40481 first

**Comment by pcorpet** [comment]
> Just in case, I don't think we want to bump `@mui/material` minimum version just for this improvement. I'm not sure if using an older version of `@mui/material` and a version of `@mui/x-data-grid` with your improvement will cause a TS error. If it does, we should add a type casting IMHO.

Thanks for the heads' up. I've added a cast and left some documentation around it.

File: packages/grid/x-data-grid/src/models/props/DataGridProps.ts [hunk 1]
```diff
@@ -304,7 +304,7 @@ export interface DataGridPropsWithDefaultValues {
    * Select the pageSize dynamically using the component UI.
    * @default [25, 50, 100]
    */
-  pageSizeOptions: Array<number | { value: number; label: string }>;
+  pageSizeOptions: readonly (number | { value: number; label: string })[];
```
Comment (oliviertassinari) [inline packages/grid/x-data-grid/src/models/props/DataGridProps.ts:307]: More consistent in the codebase?
```suggestion
  pageSizeOptions: ReadonlyArray<number | { value: number; label: string }>;
```

**Review by cherniavskii — APPROVED** [review]
Thank you!

**Pull request merged**